libs/autotagging-commons/src/main/java/org/opensearch/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.rule;
+
+import org.opensearch.autotagging.Attribute;
+import org.opensearch.autotagging.FeatureType;
+import org.opensearch.autotagging.Rule;
+import org.opensearch.rule.attribute_extractor.AttributeExtractor;
+import org.opensearch.rule.storage.AttributeValueStore;
+import org.opensearch.rule.storage.AttributeValueStoreFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * This class is responsible for managing in-memory view of Rules and Find matching Rule for the request
+ * Each auto-tagging feature should use a separate instance of this class as this avoid potential concurrency overhead
+ * in case of dynamic updates and attribute sharing scenarios
+ */
+public class InMemoryRuleProcessingService {
+
+    private AttributeValueStoreFactory attributeValueStoreFactory;
+
+    /**
+     *  Make the instance accessible using static factory method only
+     */
+    private InMemoryRuleProcessingService() {}
+
+    /**
+     * This factory methods creates an instance of InMemoryRuleProcessingService to use auto-tagging framework
+     * @param featureType
+     * @param attributeValueStoreSupplier
+     */
+    public static InMemoryRuleProcessingService createForFeature(
+        FeatureType featureType,
+        Supplier<AttributeValueStore<String, String>> attributeValueStoreSupplier
+    ) {
+        InMemoryRuleProcessingService service = new InMemoryRuleProcessingService();
+        service.attributeValueStoreFactory = AttributeValueStoreFactory.create(featureType, attributeValueStoreSupplier);
```

Comment:
If we want to create `InMemoryRuleProcessingService` per featureType, why not just take these params in a constructor?
```
InMemoryRuleProcessingService(FeatureType featureType) {
  this.attributeValueStoreFactory = new AttributesValueStoreFactory(featureType);
}
```
As anyway, we need to ensure that these classes are created only once per featureType.